d3a_interface/matching_algorithms/requirements_validators.py

```diff
@@ -0,0 +1,205 @@
+"""
+
+
+from abc import ABC, abstractmethod
+from typing import Dict, Optional, Union
+
+from d3a_interface.data_classes import Offer, Bid
+
+
+class Requirement(ABC):
+    """Interface for offer and bid requirements."""
+
+    @classmethod
+    @abstractmethod
+    def is_satisfied(
+            cls, offer: Offer, bid: Bid, requirement: Dict,
+            clearing_rate: Optional[float] = None,
+            selected_energy: Optional[float] = None) -> bool:
+        """Check whether a requirement is satisfied."""
+
+
+class TradingPartnersRequirement(Requirement):
+    """Check if trading_partners requirement is satisfied for both bid and offer."""
+
+    @classmethod
+    def is_satisfied(cls, offer: Offer, bid: Bid, requirement: Dict,
+                     clearing_rate: Optional[float] = None,
+                     selected_energy: Optional[float] = None) -> bool:
+        trading_partners = requirement.get("trading_partners", [])
+        assert isinstance(trading_partners, list),\
+            f"Invalid data type for trading partner {requirement}"
+        if trading_partners and not (
+                bid.buyer_id in trading_partners or
+                bid.buyer_origin_id in trading_partners or
+                offer.seller_id in trading_partners or
+                offer.seller_origin_id in trading_partners):
+            return False
+        return True
+
+
+class EnergyTypeRequirement(Requirement):
+    """Check if energy_type requirement of bid is satisfied."""
+
+    @classmethod
+    def is_satisfied(cls, offer: Offer, bid: Bid, requirement: Dict,
+                     clearing_rate: Optional[float] = None,
+                     selected_energy: Optional[float] = None) -> bool:
+        bid_required_energy_types = requirement.get("energy_type")
+        assert isinstance(bid_required_energy_types, list), \
+            f"Invalid data type for energy_type {requirement}"
+        offer_energy_type = (
+                offer.attributes or {}).get("energy_type", None)
+        # bid_required_energy_types is None or it is defined & includes offer_energy_type -> true
+        return not bid_required_energy_types or offer_energy_type in bid_required_energy_types
+
+
+class SelectedEnergyRequirement(Requirement):
+    """Check if energy (selected energy) requirement of bid is satisfied."""
+
+    @classmethod
+    def is_satisfied(cls, offer: Offer, bid: Bid, requirement: Dict,
+                     clearing_rate: Optional[float] = None,
+                     selected_energy: Optional[float] = None) -> bool:
+        bid_required_energy = requirement.get("energy")
+        assert isinstance(bid_required_energy, (int, float)), \
+            f"Invalid data type for energy {requirement}"
+        assert isinstance(selected_energy, (int, float)), \
+            f"Invalid data type for selected_energy {selected_energy}"
+        # bid_required_energy is None or it is defined & equals selected_energy -> true
```

Comment:
If `bid_required_energy` is `None` the method will raise an `AssertionError`. Should this comment be updated?

Reply:
Thanks for spotting that, a None value should raise an assertion error so I will adapt it accordingly